File: src/main/java/com/fast/backend/vehicle/domain/VehicleStatus.java
package com.fast.backend.vehicle.domain;

/**
 * 차량 상태 enum. Isaac 현재 계약의 HOLDING까지 보존한다.
 *
 * <p><b>확정 이전과의 차이</b>: 이전 버전은 {@code UNKNOWN/IDLE/ACTIVE/ERROR/OFFLINE} 5종만 두고
 * {@code MOVING}을 {@code ACTIVE}로 강제 변환했다. 팀 확정에 따라 <b>MOVING을 별도 상태로 보존</b>하고
 * {@code LIFTING/LOADING/UNLOADING/ESTOP}을 독립 상태로 추가했다 — 더 이상 어떤 상태도 다른 상태로
 * 흡수되지 않는다.
 *
 * <p>각 상태의 의미:
 * <ul>
 *   <li>{@link #UNKNOWN} — 상태 확인 불가</li>
 *   <li>{@link #IDLE} — 대기 또는 정지</li>
 *   <li>{@link #ACTIVE} — 일반 작업 활성 상태</li>
 *   <li>{@link #MOVING} — 주행 중</li>
 *   <li>{@link #LIFTING} — 포크 승강 중</li>
 *   <li>{@link #LOADING} — 적재 중</li>
 *   <li>{@link #UNLOADING} — 하역 중</li>
 *   <li>{@link #HOLDING} — 관제에 의한 일시 정지</li>
 *   <li>{@link #ESTOP} — 비상 정지 상태</li>
 *   <li>{@link #ERROR} — 오류 발생</li>
 *   <li>{@link #OFFLINE} — 통신 단절 또는 접속 종료</li>
 * </ul>
 *
 * <p>enum 상수 <b>선언 순서가 곧 상태 집계 응답의 항목 순서</b>다
 * ({@code VehicleService#countByStatus}가 {@code values()}를 그대로 순회한다).
 */
public enum VehicleStatus {
    UNKNOWN,
    IDLE,
    ACTIVE,
    MOVING,
    LIFTING,
    LOADING,
    UNLOADING,
    HOLDING,
    ESTOP,
    ERROR,
    OFFLINE;

    /**
     * 외부(REST 요청, MQTT 메시지)에서 들어온 원시 문자열을 안전하게 매핑한다.
     * null·빈 문자열·정의되지 않은 값은 모두 {@link #UNKNOWN}으로 처리해, 잘못된 상태값 하나 때문에
     * 차량 상태 갱신 전체가 예외로 실패하지 않도록 한다.
     *
     * <p><b>어떤 값도 다른 값으로 치환하지 않는다.</b> 특히 {@code "MOVING"}은 이제
     * {@link #MOVING} 그대로 보존된다(과거의 {@code MOVING → ACTIVE} 정규화는 prompt32.md 확정으로
     * 폐지됐다). 대소문자와 앞뒤 공백만 허용 오차로 흡수한다.
     */
    public static VehicleStatus fromRaw(String raw) {
        if (raw == null || raw.isBlank()) {
            return UNKNOWN;
        }
        try {
            return VehicleStatus.valueOf(raw.trim().toUpperCase());
        } catch (IllegalArgumentException e) {
            return UNKNOWN;
        }
    }
}

File: frontend/monitoring-control-system/lib/vehicleStatus.ts
import type { VehicleStatus } from "@/types/monitoring"

/**
 * 백엔드 VehicleStatus enum과 동일한 목록.
 * 배열 순서는 백엔드 enum 선언 순서와 동일하다.
 */
export const VEHICLE_STATUS_VALUES = [
  "UNKNOWN",
  "IDLE",
  "ACTIVE",
  "MOVING",
  "LIFTING",
  "LOADING",
  "UNLOADING",
  "HOLDING",
  "ESTOP",
  "ERROR",
  "OFFLINE",
] as const satisfies readonly VehicleStatus[]

const VEHICLE_STATUS_SET: ReadonlySet<string> = new Set(VEHICLE_STATUS_VALUES)

/**
 * 임의의 값을 안전하게 VehicleStatus 로 정규화한다.
 *
 * Isaac 경로의 상태 payload 는 백엔드에서 정규화되지 않은 <b>원본 문자열</b>이 그대로 오기 때문에
 * (IsaacVehicleStatusEventData.status) 프론트가 방어적으로 처리해야 한다.
 *
 * - 문자열이 아니면 UNKNOWN
 * - 앞뒤 공백 제거 후 대문자로 변환
 * - 허용값이 아니면 UNKNOWN (알 수 없는 값을 임의의 다른 상태로 흡수하지 않는다)
 */
export function normalizeVehicleStatus(value: unknown): VehicleStatus {
  if (typeof value !== "string") {
    return "UNKNOWN"
  }
  const normalized = value.trim().toUpperCase()
  return VEHICLE_STATUS_SET.has(normalized) ? (normalized as VehicleStatus) : "UNKNOWN"
}

/**
 * 위치/telemetry 이벤트가 보고한 상태를 기존 화면 상태에 병합할 수 있는 값으로 바꾼다.
 * 계약 밖 문자열과 UNKNOWN 은 "갱신 정보 없음"인 null 로 반환해, REST 로 이미 알고 있던
 * ERROR/IDLE 같은 상태를 불완전한 위치 이벤트가 지우지 않게 한다.
 */
export function normalizeReportedVehicleStatus(value: unknown): VehicleStatus | null {
  const normalized = normalizeVehicleStatus(value)
  return normalized === "UNKNOWN" ? null : normalized
}

/** 차량 ID에서 미니맵 마커용 짧은 라벨을 만든다. 예: "SIM-F01" → "F01" */
export function toShortLabel(vehicleId: string): string {
  const trimmed = vehicleId.trim()
  if (!trimmed) {
    return "?"
  }
  const lastSegment = trimmed.split("-").pop()
  return (lastSegment && lastSegment.length > 0 ? lastSegment : trimmed).slice(0, 6)
}
